ts/nni_manager/core/nniExperimentsManager.ts

```diff
@@ -0,0 +1,160 @@
+import { Deferred } from 'ts-deferred';
+
+interface CrashedInfo {
+    experimentId: string;
+    isCrashed: boolean;
+}
+
+interface FileInfo {
+    buffer: Buffer;
+    mtime: number;
+}
+
+class NNIExperimentsManager implements ExperimentManager {
+    private experimentsPath: string;
+    private log: Logger;
+    private profileUpdateTimer: {[key: string]: any};
+
+    constructor() {
+        this.experimentsPath = getExperimentsInfoPath();
+        this.log = getLogger();
+        this.profileUpdateTimer = {};
+    }
+
+    public async getExperimentsInfo(): Promise<JSON> {
+        const fileInfo = await this.readExperimentsInfo();
+        const experimentsInformation = JSON.parse(fileInfo.buffer.toString());
+        const expIdList: Array<string> = Object.keys(experimentsInformation).filter((expId) => {
+            return experimentsInformation[expId]['status'] !== 'STOPPED';
+        });
+        const updateList: Array<CrashedInfo> = (await Promise.all(expIdList.map((expId) => {
+            return this.checkCrashed(expId, experimentsInformation[expId]['pid']);
+        }))).filter(crashedInfo => crashedInfo.isCrashed);
+        if (updateList.length > 0){
+            const result = await this.withLock(this.updateAllStatus, updateList.map(crashedInfo => crashedInfo.experimentId), fileInfo.mtime);
+            if (result !== undefined) {
+                return result;
+            } else {
+                return this.getExperimentsInfo();
+            }
+        } else {
+            return JSON.parse(JSON.stringify(Object.keys(experimentsInformation).map(key=>experimentsInformation[key])));
+        }
+    }
+
+    public setExperimentPath(newPath: string): void {
+        if (newPath[0] === '~') {
+            newPath = path.join(os.homedir(), newPath.slice(1));
+        }
+        if (!path.isAbsolute(newPath)) {
+            newPath = path.resolve(newPath);
+        }
+        this.log.info(`Set new experiment information path: ${newPath}`);
+        this.experimentsPath = newPath;
+    }
+
+    public setExperimentInfo(experimentId: string, key: string, value: any): void {
+        try {
+            if (this.profileUpdateTimer[key] !== undefined) {
+                clearTimeout(this.profileUpdateTimer[key]);
+                this.profileUpdateTimer[key] = undefined;
+            }
+            this.withLockSync(() => {
+                const experimentsInformation = JSON.parse(fs.readFileSync(this.experimentsPath).toString());
+                if (experimentsInformation[experimentId]) {
+                    experimentsInformation[experimentId][key] = value;
+                    fs.writeFileSync(this.experimentsPath, JSON.stringify(experimentsInformation));
+                } else {
+                    this.log.error(`Experiment Manager: Experiment Id ${experimentId} not found, this should not happen`);
+                }
+            });
+        } catch (err) {
+            this.log.error(err);
```

Comment:
is this err recoverable?

Reply:
This error is due to other processes lock the file, I add some details to distinguish this kind of error.